### PR TITLE
Add SM100 H16 DSA backward specialization

### PR DESCRIPTION
**Pull request by icavan**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` on every changed file and committed any formatting changes.
- [ ] Labels requested: `cat-enhancements`, `mod-cutedsl`, and `orig-external` (the external contributor account does not have permission to add labels to the NVIDIA repository).

## Affected area

FE OSS kernels / CuTeDSL, DSA backward performance, tests, documentation, and benchmark discoverability.

## Summary

This PR adds a dedicated SM100 DeepSeek Sparse Attention backward specialization for:

- `H = 16`
- `d_qk = 576`, `d_v = 512`
- BF16 or FP16 inputs/outputs with FP32 accumulation
- an M128 selected-KV tile

The public API is unchanged. The SM100 dispatcher selects the new backend only for H16/D576; H16/D512 and all other head counts continue to use the existing generic M64 backend.

The specialization keeps the same high-level DSA backward algorithm while reorganizing the main kernel around a KV-major M128 × N16 score plane:

- 16 indexed KV-loader warps gather one 128-row sparse KV tile with `cp.async`;
- four compute warps produce probability and softmax-gradient operands;
- one UMMA issuer keeps five persistent FP32 dQ fragments in TMEM;
- eight reducer warps drain FP32 dKV fragments into the existing global workspace;
- fenced TMEM-to-register reads release aliased TMEM before vector global reductions finish;
- the 144 KiB K/V SMEM allocation is reused as the dQ epilogue staging buffer after K is dead.

The full operator remains a four-kernel sequence: statistics preprocessing, the H16/M128 main backward kernel, FP32-to-output-dtype dKV conversion, and dSink reduction.

### Source-audited memory and pipeline diagram

The diagram below documents the physical 188 KiB SMEM layout, 512-column TMEM map, warp ownership, lifetime reuse, mathematical shapes, and the auxiliary-kernel boundary. It is documentation hosted outside this repository and is not part of the PR diff.

SM100 H16 M128 DSA backward memory and pipeline

## Why

For H16, the generic M64 mapping leaves half of the M128 UMMA row capacity unused and requires twice as many sparse-KV tile iterations. Mapping the 128 selected KV rows to UMMA M and the real 16 heads to N uses the full M datapath without padding the head dimension to H64.

The implementation is intentionally kept in a separate H16 kernel class so that the existing H32/H64 generic path retains its current code generation, register allocation, and dispatch behavior.

## Performance

Measured through the public `cudnn.DSA.sparse_attention_backward_wrapper` API on an NVIDIA GB200 (SM100), CUDA 13.0, cuDNN 9.15.1, PyTorch `2.8.0a0+34c6371d24.nv25.08`, and CUTLASS DSL 4.5.2.

Common shape: `S_q = S_kv = 4096`, H16, DQK576/DV512, BF16, attention sink enabled, full `topk_length`. CUDA events cover 100 iterations after 20 warmups. The timed region includes buffer zeroing, workspace allocation, statistics preprocessing, the main kernel, dKV conversion, and dSink. The baseline forces the existing generic M64 backend from the same source revision.

| top-k | Generic M64 | H16 M128 | Speedup | Latency reduction |
|---:|---:|---:|---:|---:|
| 512 | 1.933 ms | 1.041 ms | 1.857× | 46.1% |
| 1024 | 3.645 ms | 1.867 ms | 1.952× | 48.8% |
| 2048 | 7.064 ms | 3.522 ms | 2.006× | 50.1% |

## Related issues

None.

## API and compatibility impact

- No public API or argument changes.
- New automatic dispatch only for SM100 H16/D576/DV512.
- H16/D512 remains on the tested generic M64 fallback.
- Other head counts and SM90 behavior are unchanged.
- dQ is numerically aligned with the generic implementation; M128 changes its reduction grouping, so bitwise identity with generic M64 is not promised.
- dKV continues to use cross-CTA FP32 atomic accumulation and is therefore validated numerically rather than for run-to-run bitwise determinism.

### Formatting and static checks on the submitted commit

```text
pre-commit run --files <all six changed files>
clang-format: skipped (no C/C++ files)
black: passed
black-jupyter: passed

git diff --check origin/develop..HEAD: passed
Python AST parsing for all changed Python files: passed
```

### SM100 correctness

The DSA backward L0 test file was run on GB200 with reduced sequence sizes:

```text
26 passed, 1 skipped, 2 warnings in 86.09s
```

Coverage includes H16/D576 auto-dispatch, H16/D512 fallback, BF16 and FP16, with and without `topk_length`, zero-length rows, boundary and tail top-k lengths 127/128/129 and 511/512/513, non-default streams, non-contiguous auxiliary inputs, and API contract validation.

### Compute Sanitizer

Sanitizer shape: `S_q=2`, `S_kv=640`, `topk=513`, H16, DQK576/DV512, BF16, without `topk_length`. This exercises five M128 iterations, TMEM/SMEM reuse, and a final one-row sparse tail.

```text
compute-sanitizer --tool memcheck --error-exitcode 99 \
  --kernel-name kns=dsa_bwd_sm100_h16 \
  env PYTHONPATH=$REPO/python python -m pytest -q \
  fe_api/dsa/test_DSA_sparse_attention_backward.py::test_DSA_sparse_attention_backward_sm100_h16_partial_max_topk

1 passed, 2 warnings in 7.33s
ERROR SUMMARY: 0 errors
```

```text
compute-sanitizer --tool racecheck --racecheck-report all \
  --error-exitcode 99 --kernel-name kns=dsa_bwd_sm100_h16 \
  env PYTHONPATH=$REPO/python python -m pytest -q \
  fe_api/dsa/test_DSA_sparse_attention_backward.py::test_DSA_sparse_attention_backward_sm100_h16_partial_max_topk

1 passed, 2 warnings in 7.45s
RACECHECK SUMMARY: 0 hazards displayed (0 errors, 0 warnings)
```

## Summary by CodeRabbit

* **New Features**
  * Added automatic sparse-attention backward kernel selection for SM100 hardware.
  * Added optimized support for 16 query heads with a head dimension of 576.
  * Removed the need to provide backend or tile-size arguments for supported configurations.

* **Bug Fixes**
  * Improved handling of zero-length and partial sparse tiles in backward attention.

* **Tests**
  * Expanded coverage for SM100 dispatch, numerical accuracy, boundary cases, and the new 16-head configuration.

* **Documentation**
  * Documented SM100 kernel selection behavior and added a benchmark example.

### CODE REVIEW

**Review by Anerudhan — APPROVED** [review]

**Comment by Anerudhan** [comment]
@cudnn-ci-bot run oss

**Comment by Anerudhan** [comment]
Thanks for the PR. Merging
